File: practice-projects/gin-ddd/internal/infrastructure/persistence/user/user_converter.go
package user

import (
	"database/sql"

	userModel "gin-ddd/internal/domain/model/user"
)

// toDO 将用户领域模型转换为数据对象。
func toDO(u *userModel.User) *UserDO {
	if u == nil {
		return nil
	}
	return &UserDO{
		ID:          u.ID,
		Name:        u.Name,
		Email:       u.Email,
		Phone:       nullStringFromPtr(u.Phone),
		Address:     nullStringFromPtr(u.Address),
		CreatedTime: u.CreatedTime,
		UpdatedTime: u.UpdatedTime,
	}
}

// toModel 将数据对象还原为领域模型,经聚合根 Restore 工厂构建。
func toModel(do *UserDO) *userModel.User {
	if do == nil {
		return nil
	}
	return userModel.Restore(
		do.ID,
		do.Name,
		do.Email,
		ptrFromNullString(do.Phone),
		ptrFromNullString(do.Address),
		do.CreatedTime,
		do.UpdatedTime,
	)
}

// toModels 批量转换。
func toModels(dos []*UserDO) []*userModel.User {
	if dos == nil {
		return nil
	}
	out := make([]*userModel.User, 0, len(dos))
	for _, do := range dos {
		out = append(out, toModel(do))
	}
	return out
}

func nullStringFromPtr(p *string) sql.NullString {
	if p == nil {
		return sql.NullString{Valid: false}
	}
	return sql.NullString{String: *p, Valid: true}
}

func ptrFromNullString(ns sql.NullString) *string {
	if !ns.Valid {
		return nil
	}
	s := ns.String
	return &s
}

File: practice-projects/springboot4ddd/src/main/java/com/github/microwind/springboot4ddd/infrastructure/repository/user/UserConverter.java
package com.github.microwind.springboot4ddd.infrastructure.repository.user;

import com.github.microwind.springboot4ddd.domain.model.user.User;

import java.util.List;
import java.util.stream.Collectors;

/**
 * 用户领域模型 ↔ 数据对象 显式转换器
 *
 * <p>放在 infrastructure 层，避免领域模型反向依赖持久化结构。
 * 所有字段映射在编译期可见，字段调整时编译器会立刻报错。
 *
 * @author jarry
 * @since 1.0.0
 */
public final class UserConverter {

    private UserConverter() {
    }

    public static UserDO toDO(User user) {
        if (user == null) {
            return null;
        }
        return UserDO.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .phone(user.getPhone())
                .wechat(user.getWechat())
                .address(user.getAddress())
                .createdTime(user.getCreatedTime())
                .updatedTime(user.getUpdatedTime())
                .build();
    }

    public static User toModel(UserDO userDO) {
        if (userDO == null) {
            return null;
        }
        return User.restore(
                userDO.getId(),
                userDO.getName(),
                userDO.getEmail(),
                userDO.getPhone(),
                userDO.getWechat(),
                userDO.getAddress(),
                userDO.getCreatedTime(),
                userDO.getUpdatedTime()
        );
    }

    public static List<User> toModelList(List<UserDO> dos) {
        if (dos == null) {
            return null;
        }
        return dos.stream()
                .map(UserConverter::toModel)
                .collect(Collectors.toList());
    }
}
